docs/index_while_building.md

```diff
@@ -0,0 +1,102 @@
+# Index While Building in Bazel
+
+This document summarizes the status of index while building in Bazel and the
+plan to improve it. "Index while building" encompasses the writing of index data
+from swift and clang and the usage of that data in Xcode.
+For a high level design of the feature in LLVM, check the paper [Adding Index-While-Building support to Clang](https://docs.google.com/document/d/1cH2sTpgSnJZCkZtJl1aY-rzy4uGPcrI-6RrUpdATO2Q/edit)
+
+## High level summary and status
+
+At the time of writing there's prior art that adds `-index-store-path` to
+`rules_swift` and `rules_ios` but enabling these code paths slows down build
+time in testing.
+
+`swift` and `clang` write index data based on the status of what resides in
+index-store-path. When using a transient per-swift-library index, it writes O(
+M imports * N libs) indexer data and slows down compilation significantly: to
+the tune of 300% slow down and 6GB+ index data in benchmarks. `rules_swift`
+likes to use per `swift_library` data in order to remote cache indexes.
+`rules_ios` uses a per `apple_library` index to reduce `index-import` wallclock
+time.  Adding "Index while building" to Bazel needs special consideration to
+both preserve performance characteristics of the original architecture of a
+global index and interoperate with Bazel's remote execution and caching.
+
```

Comment:
It might be useful to spell out how this design saves so much bandwidth on the cache VS current and the special sauce in https://github.com/lyft/index-import/pull/53 to computing _what_ units and records to cache.